networkx/algorithms/similarity.py

```diff
@@ -1642,6 +1749,182 @@ def panther_similarity(
+        accuracy at the cost of more computation. Defaults to 0.5 as recommended
+        in [1]_.
+    delta : float
+        The probability that $S$ is not an epsilon-approximation to (R, phi)
+    eps : float
+        The error bound for similarity approximation. This controls the accuracy
+        of the sampled paths in representing the true similarity. Smaller values
+        yield more accurate results but require more sample paths. If None, a
+        value of ``sqrt(1/|E|)`` is used, which the authors found empirically
+        effective.
+    weight : string or None, optional (default="weight")
+        The name of an edge attribute that holds the numerical value
+        used as a weight. If None then each edge has weight 1.
+
+    Returns
+    -------
+    similarity : dict
+        Dict of nodes to similarity scores (as floats).
+        Note: the self-similarity (i.e., `node`) is not included in the dict.
+
+    Examples
+    --------
+
+        >>> import networkx as nx
+        >>> G = nx.star_graph(100)
+        >>> sim = nx.panther_vector_similarity(G, 0)
+
+    References
+    ----------
+    .. [1] Zhang, J., Tang, J., Ma, C., Tong, H., Jing, Y., & Li, J.
+           Panther: Fast top-k similarity search on large networks.
+           In Proceedings of the ACM SIGKDD International Conference
+           on Knowledge Discovery and Data Mining (Vol. 2015-August, pp. 1445–1454).
+           Association for Computing Machinery. https://doi.org/10.1145/2783258.2783267.
+    """
+    import numpy as np
+    from scipy.spatial import cKDTree as KDTree
+
+    # Use helper method to prepare common data structures but keep isolates in the graph
+    paths = _prepare_panther_paths(
+        G,
+        source,
+        path_length=path_length,
+        c=c,
+        delta=delta,
+        eps=eps,
+        weight=weight,
+        remove_isolates=False,
+        k=k,
+    )
+
+    G = paths.G
+    inv_node_map = paths.inv_node_map
+    node_map = paths.node_map
+    index_map = paths.index_map
+    inv_sample_size = paths.inv_sample_size
+    eps = paths.eps
+    num_nodes = G.number_of_nodes()
+
+    # Ensure D doesn't exceed the number of nodes
+    if num_nodes < D:
+        raise nx.NetworkXUnfeasible(
+            f"The number of requested similarity scores {D} is greater than the number of nodes {num_nodes}."
+        )
+
+    # We'll use heaps to track the top D similarity scores for each node
+    # This avoids allocating the full n×n similarity matrix
+    theta = np.zeros((num_nodes, D))
+
+    # Pre-convert index_map values to sets for faster intersection
+    index_map_sets = {node: set(paths) for node, paths in index_map.items()}
+
+    # Calculate the path similarities for each node
+    for vi_idx, vi in enumerate(G.nodes):
+        # Get the set of paths for the current node
+        vi_paths = index_map_sets[vi]
+
+        # Use a min heap to track the top D scores
```

Comment:
I think this can be made faster using quickselect/introselect, but I don't know if the added code complexity is worth the reduction in time complexity (O(n + D log D) instead of O(n log D)). What are typical values of n and D?

Reply:
This is out of scope for this PR and might not even be faster in practice.

Reply:
My understanding is the typical number of nodes is 100k+ or millions, whereas k and D are usually 50-100.